### PR TITLE
Badges

### PR DESCRIPTION
**Pull request by hyper2snyper**
## About The Pull Request
New thing: The 7E Chameleon Badge.
A new attachment for uniforms/armor/berets and helmets. 
It is a colorable badge that comes in many different styles (Changeable by Alt - Click)
Colored by being attacked by green facepaint. 
Available in the surplus uniform vendor at that bottom of the uniform tab.

## Why It's Good For The Game
Customization content

## Changelog
:cl:
add: New thing: The 7E Chameleon Badge. Colorable Badge that can be found in the surplus uniform vendor. Recolorable by facepaint and restyleable by Alt Click. Can attach to armor, berets, uniform and helmets.
/:cl:

### CODE REVIEW

**Review by Kuro020 — APPROVED** [review]

**Comment by InterroLouis** [comment]
Badges?! We don't need no stinkin' badges!

File: code/datums/greyscale/greyscale_configs.dm [hunk 1]
```diff
@@ -167,3 +167,181 @@
...
+	json_config = 'code/datums/greyscale/json_configs/badge/badge_beret_side.json'
+
+/datum/greyscale_config/badge/berets/triangle
+	icon_file = 'icons/mob/modular/badge/triangle.dmi'
+
+/datum/greyscale_config/badge/berets/circle
+	icon_file = 'icons/mob/modular/badge/circle.dmi'
+
+/datum/greyscale_config/badge/berets/square
+	icon_file = 'icons/mob/modular/badge/square.dmi'
```
Comment (TiviPlus) [inline code/datums/greyscale/greyscale_configs.dm:294]: You dont have to make new icons ffor all of these you can just change the icon state

File: code/modules/clothing/clothing.dm [hunk 2]
```diff
@@ -4,6 +4,16 @@
...
+	///Assoc list of available slots.
+	var/list/attachments_by_slot = list()
+	///Typepath list of allowed attachment types.
+	var/list/attachments_allowed = list()
+
+	///Pixel offsets for specific attachment slots. Is not used currently.
+	var/list/attachment_offsets = list()
+	///List of attachment types that is attached to the object on initialize.
+	var/list/starting_attachments = list()
+
```
Comment (TiviPlus) [inline code/modules/clothing/clothing.dm:16]: lazylist + string list this?

File: code/modules/clothing/modular_armor/attachments/badges.dm [hunk 3]
```diff
@@ -0,0 +1,111 @@
+/obj/item/armor_module/armor/badge
+	name = "E7 Chameleon Badge"
+	desc = "The E7 Chameloen Badge uses brand new and revolutionary technology to make your gear look even cooler. It is capable of changing into a variety of different shapes (Alt-Click), changing colors (Attack with Green Facepaint), and attaching to nearly all clothing, helmets, berets, and or armor!"
+	greyscale_config = /datum/greyscale_config/badge/lchest
+	greyscale_colors = "#afafad"
+	slot = ATTACHMENT_SLOT_BADGE
+	flags_attach_features = ATTACH_REMOVABLE|ATTACH_SAME_ICON|ATTACH_APPLY_ON_MOB|ATTACH_NO_HANDS
+	///List of selectable greyscale configs for wearing the badge at different levels and on different objects.
+	var/list/style_list = list(
+		"Shield (Left Chest)" = /datum/greyscale_config/badge/lchest,
```
Comment (TiviPlus) [inline code/modules/clothing/modular_armor/attachments/badges.dm:10]: string or static this

File: code/modules/clothing/modular_armor/attachments/badges.dm [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+/obj/item/armor_module/armor/can_attach(obj/item/attaching_to, mob/user)
+	. = ..()
+	var/allowed = TRUE
+	if((istype(attaching_to, /obj/item/clothing/under) || istype(attaching_to, /obj/item/clothing/suit)) && (!ispath(greyscale_config, /datum/greyscale_config/badge/lchest) && !ispath(greyscale_config, /datum/greyscale_config/badge/rchest) && !ispath(greyscale_config, /datum/greyscale_config/badge/larm) && !ispath(greyscale_config, /datum/greyscale_config/badge/rarm)))
+		allowed = FALSE
+	if((istype(attaching_to, /obj/item/clothing/head/beret) || istype(attaching_to, /obj/item/clothing/head/tgmcberet)) && (!ispath(greyscale_config, /datum/greyscale_config/badge/beretf) && !ispath(greyscale_config, /datum/greyscale_config/badge/berets)))
+		allowed = FALSE
+	if((istype(attaching_to, /obj/item/clothing/head) && !istype(attaching_to, /obj/item/clothing/head/beret && !istype(attaching_to, /obj/item/clothing/head/tgmcberet))) && (!ispath(greyscale_config, /datum/greyscale_config/badge/lhelm) && !ispath(greyscale_config, /datum/greyscale_config/badge/rhelm)))
+		allowed = FALSE
```
Comment (TiviPlus) [inline code/modules/clothing/modular_armor/attachments/badges.dm:91]: wtf is this hell check

File: code/modules/clothing/modular_armor/modular.dm [hunk 5]
```diff
@@ -367,10 +349,9 @@
 	///Current varient selected.
 	var/current_variant
 
-/obj/item/clothing/head/modular/Initialize(mapload)
+/obj/item/clothing/head/modular/Initialize()
 	. = ..()
-	AddComponent(/datum/component/attachment_handler, attachments_by_slot, attachments_allowed, attachment_offsets, starting_attachments, null, null, null, attachment_overlays)
-	update_icon()
+	update_icon() //Update for greyscale.
```
Comment (TiviPlus) [inline code/modules/clothing/modular_armor/modular.dm:354]: update icon doesnt call greyscale or am i missing something

File: code/modules/clothing/modular_armor/modular.dm [hunk 6]
```diff
@@ -657,6 +648,8 @@
 		/obj/item/armor_module/module/antenna,
 
 		/obj/item/armor_module/storage/helmet,
+
+		/obj/item/armor_module/armor/badge,
```
Comment (TiviPlus) [inline code/modules/clothing/modular_armor/modular.dm:652]: random spaces throughoout

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/datums/greyscale/greyscale_configs.dm (hunk 1 above)
Comment (hyper2snyper) [reply code/datums/greyscale/greyscale_configs.dm:294]: ok we'll have to talk about greyscale. You obviously know something I dont.

File: code/modules/clothing/modular_armor/attachments/badges.dm (hunk 4 above)
Comment (hyper2snyper) [reply code/modules/clothing/modular_armor/attachments/badges.dm:91]: Its checking the currently selected sytle (aka greyscale datum) and checking it against the type you are attaching it to. No beret style badges attached to uniforms

File: code/modules/clothing/modular_armor/modular.dm (hunk 5 above)
Comment (hyper2snyper) [reply code/modules/clothing/modular_armor/modular.dm:354]: Without the update_icon(), the thing was initializing invisible.

**Comment by SpaceLoveSs13** [comment]
Badges don't work on assault helmets by the way @hyper2snyper and gives the following message:-
 
image

**Pull request merged**